### ISSUE TITLE
H spacing on the period when used in a between two numbers context

### ISSUE THREAD

**Issue by chrissimpkins** [googlefonts/googlesans-flex#1046]
This is a request from Ryan Kiley (Material UX).  Let's discuss the current default spacing on the period and whether this is most appropriate as a default spacing change, or a new calt feature when the period is directly between two numbers.  He is attempting to support millisecond times in clocks in the example attached below.  

syntax: `NUMBER-COLON-NUMBER-PERIOD-NUMBER`

Are there any downsides to this spacing change, including when it only shows up between numbers in a calt feature?

Ryan request:

Can the `.` can have the same padding as the `:` in GSF? Maybe something conditional, like between two numbers? like how the colon becomes a "time separator" with the raised baseline?

**Comment by davelab6** [comment]
Conditional sounds good to me, same as the colon vertical position conditional

**Comment by octaviopardo** [comment]
so right now in the font there is period, period.tf, colon, colon.tf, and colon.time; 

colon, period and colon.time have all the same metrics. 
colon.tf, period.tf have the same metrics

So the issue in the video? they are switching in between colon.time and period.tf. The tf feature is probably activated so all the figures (and in this case the comma and the period) have the same metrics. If they want to use the period with the same width as the colon.time, they need to deactivate the tf feature for that glyph

**Comment by MariannaPaszkowska** [comment]
Hi @octaviopardo I got curious about something else and I double-checked. I can reproduce the issue, and I think it's a bug in the font. I see a slight difference in the spacing between colon.time and period in 18opsz, 25 condensed and 400 weight:

Propotional spacing (colon.time and period):
<img width="854" alt="Screenshot 2024-09-24 at 2 43 49 PM" src="https://github.com/user-attachments/assets/30381dfe-94af-4e11-8e4b-cc211e646c3a">


Tabular settings (colon.time and period):
<img width="873" alt="Screenshot 2024-09-24 at 2 45 07 PM" src="https://github.com/user-attachments/assets/7364df60-53e2-485a-b1f9-3cec5e7b8bfc">


BTW, do you think it would make sense to add colon.time.tf?

**Comment by MariannaPaszkowska** [comment]
@octaviopardo The difference is caused because of kerning. If kerning if off in Figma spacing is equal.

**Comment by MariannaPaszkowska** [comment]
@chrissimpkins So I think that the user could deactivate kerning for that text box, or we could add period.time which is not kerned... but it would mean a bigger file-size.

**Comment by octaviopardo** [comment]
exactly, I was going to say that. In that case the difference in between both is the caused by kerning. Since the whole clock thinggy uses either tf symbols or the colon time, I would advise to deactivate the kerning

**Comment by MariannaPaszkowska** [comment]
But I think we should refine the .tf feature: the colon.time should get replaced by colon.tf when the .tf feature is on. Currently, when .tf is activated on a text box where the colon is between numbers, the colon.time stays contextually replaced and as a result the colon looks tightly spaced.

**Comment by octaviopardo** [comment]
But I think that s how it s supossed to work, isn't?
I think by doing what @MariannaPaszkowska  suggest you are over complicating the task of the designers handling Figma, as you force them to select individual glyphs inside the text box to behave as they want for the clock (either they select the box, they select tf feature and then individually they have to deactivate the feature in the colon; or they activate the colon.time feature, and individually select the numbers to activate the tabular feature). 

 I think for the clock people it makes more sense to activate both features and therefore having the tabular figures plus the colon.time superseeding colon.tf. The issue you showed in your comment should not occur whenever they use the features correctly, because the period has no kerning with tabular figures.

In my opinion all this issue is happening because there is a single glyph that is quite uncommon, the period.tf, and the designer did not understand that s what he was seeing when he changed the glyphs. But the font is fine, it just needs explanation

**Comment by octaviopardo** [comment]
we could creat an additional glyph, period.time, exactly equal to period, but that superseeds period.tf when the clock feature is active

**Comment by MariannaPaszkowska** [comment]
Actually, there’s no clock feature here. The replacement is happening in the `calt` feature, which is on by default. In this case, I think it would be worth considering adding a `period.time` character on the same width as `colon.time`.

What I meant is that right now, to access `colon.tf`, users have to deactivate the contextual feature (`calt`), which is on by default as per the OT spec. If users don’t have an easy way to turn off `calt`, it might be more convenient to make the optional `.tf` feature replace the narrow `colon.time` with `colon.tf`, since it's already vertically aligned with the figures anyway...

That said, you’re right—having the `colon.time` character in `calt`, that is on by default makes sense for clocks. And for users who are using `.tf` figures and need a tabular-width colon, they can just disable `calt`. It’s a good solution for designers wanting `.tf` figures in clock displays and narrow colon (the `colon.time`).

**Comment by MariannaPaszkowska** [comment]
We could also avoid adding `period.time` and update the feature code, since `period` is always on the same width as `colon.time`. So the `calt` feature could replace `period.tf` back to `period`.

or we could just educate users on switching off the `.tf` feature on the `period` glyph only in such cases.

**Comment by chrissimpkins** [comment]
Let's move this up to a v2.x release from the current v3.1 where it is milestoned.  Ryan needs this edit for the Clock application as part of the BC25 rollout next year (note: this is different than the lockscreen clock work that is happening).  Ideally we move this into the Pixel system on our Nov 20 ship to SysUI eng calendar.

Can you summarize your current thinking about the approach here?

**Comment by octaviopardo** [comment]
@chrissimpkins  
IMO 
1) Easy fix. We don t do anything. We educate users. I need a 15 min talk with Ryan to explain him the issue, why is happening and how to fix it. Risks: none. If you have not received any complain from other users I would go for this one
2) Less easy but still easy: we change the open type code so whenever calt is activated (the colon.time feature), /period.tf gets replaced by /period. This should be Ok BUT you would have to check first with everyone who is using period.tf (is there anyone??) how are they using it and if they are using it in combination with other features… like calt

Let me know what you think!

**Comment by chrissimpkins** [comment]
@ryankiley invite to repo sent.  He'll connect with you in this thread

**Comment by chrissimpkins** [comment]
> @chrissimpkins IMO
> 
> 1. Easy fix. We don t do anything. We educate users. I need a 15 min talk with Ryan to explain him the issue, why is happening and how to fix it. Risks: none. If you have not received any complain from other users I would go for this one
> 2. Less easy but still easy: we change the open type code so whenever calt is activated (the colon.time feature), /period.tf gets replaced by /period. This should be Ok BUT you would have to check first with everyone who is using period.tf (is there anyone??) how are they using it and if they are using it in combination with other features… like calt
> 
> Let me know what you think!

What are some other `[NUMBER].[NUMBER]` contexts that we can think of and do we believe that the spacing Ryan requested is detrimental in any of those use cases?  It might actually be beneficial?  e.g., decimals `1.231`, versions `v1.2`

Ignoring impact on products, what is the best type design approach for the period between two numbers?

**Comment by ryankiley** [comment]
Hi all, thanks @chrissimpkins.

I'm working on a stopwatch design. In my design I have monospaced numbers enabled. 

This is to make sure the numbers don't move around when they're in animation.
(Note: here I have worked with a motion designer to achieve the placement of the `.` I wanted)

What I noticed was that the `.` has a large space between the number—much larger than when using `:`.

This would match the behavior of San Francisco, where the `.` uses the same space as the `:`.

<img width="1002" alt="Screenshot 2024-10-02 at 11 55 33 AM" src="https://github.com/user-attachments/assets/06a68665-a5d2-4960-ae31-ba257d0e6075">

Here's the level of customization I have in Figma, as a reference:

![2024-10-02 11 58 05](https://github.com/user-attachments/assets/cfa24b88-8e92-47e4-ad7f-63d076d83b92)

**Comment by Lorp** [comment]
> Ignoring impact on products, what is the best type design approach for the period between two numbers?

I’m surprised that `tnum` turns on a monospaced period. Is Google Sans intentionally doing something novel here?

**Comment by MariannaPaszkowska** [comment]
@ryankiley Thank you for your comment, this is very helpful. I think it would make sense that we adjust the spacing of the tabular punctuation and put it on the half-width. How about we prepare test fonts and ship them to you so you can let us know if it works for you?
@Lorp it was intentional, but we are considering changing that.

**Comment by octaviopardo** [comment]
Hi @ryankiley 
Thanks a lot for sharing those cool images! You got me so excited :D 

So GSF has a quite unusual glyph –as Laurence said–, the tabular period. That means that there is a glyph, the period, that same as the numbers, has a version that when you active "tabular numbers" open type feature (the 'numbers style' row) also changes the width of the period to match it with the tabular numbers. 

So! there are two possibilities here and we would love to hear your feedback regarding what you need.
1) you could have a solution right now, by simply selecting the period in the text box and switching the numbers style row to the first position(default) or the second position (proportional figures).
2) if operatively you think this is not ideal, we could also change the code in the font, so whenever the "contextual alternates" feature is activated, you do not only get the special colon that you are getting now, but also the period that you need, the default one
3) we could also prepare what Marianna is suggesting, a version of the font where you can get this "half-monospace-figures-width" period 

Let us know if the first solution worked for you! if not, we will make any of the second or third solutions happen! :)

**Comment by ryankiley** [comment]
Option 2 sounds the best to me! Unless there are any knock-on effects of that choice you all are concerned about.

Now that I understand where this behavior comes from, I can use Option 1 in the meantime.

**Comment by ryankiley** [comment]
@chrissimpkins gave me a text file to see the spacing update. This looks good to me.

The spacing is different than the production instance of Google Sans Flex when turning tabular figures off for the `:` and `.`

The top text here is latest version of Google Sans Flex we have in Figma. The bottom is the test file:
![Image](https://github.com/user-attachments/assets/4fce0bc9-57a3-4019-a4f4-27439959d52a)

**Comment by chrissimpkins** [comment]
Good to hear that you are happy with the new colon and period spacing, Ryan!

@octaviopardo will have a look to confirm that we did not regress on the default proportional colon and period spacing.  I'm not sure why that is happening. I don't believe that it is intentional. There was no plan to adjust our proportional punctuation to address your request here.

**Comment by chrissimpkins** [comment]
@ryankiley can you share your Figma mockup where you produced this spacing?  Octavio cannot reproduce it.

**Comment by octaviopardo** [comment]
Hi @ryankiley so here is the explanation of the deviation you are seeing
for the demo font we provided you, we made the colon.tf and the period.tf half the width of the tabular figures. Both get activated automatically when you select the whole box and activate the tabular feature (in Figma is called monospace BTW). Because the two text boxes are set to "alignment -> center" the two glyphs in the middle are perfectly aligned in both lines but in the bottom line the colon and the period are slightly wider than in the top, so the four glyphs on the sides (two left and two right) are further apart from the center in the bottom line that in the top line

in the bottom line the tabular feature is not set to off in the punctuation for sure, because the "not-tabular-colon" rests in the base line, so that one is actually the tabular. Also the width of the period is too wide for the proportional.

**Comment by ryankiley** [comment]
Gotcha. I think I follow. Thanks @octaviopardo! 

LGTM

**Comment by MariannaPaszkowska** [comment]
merged in #1073

**Comment by RickyDaMa** [comment]
Reopening this, I don't think the feature code adjustment to period made it in in #1073. See https://github.com/googlefonts/pixel-brand-fallback-stack/issues/12

**Comment by MariannaPaszkowska** [comment]
I noticed that we need a new import from this branch: GSF-Ashler-v3.0-TabularPunct @RickyDaMa

**Comment by MariannaPaszkowska** [comment]
imported in #1079

**Comment by MariannaPaszkowska** [comment]
Tabular characters changed: `period.tf, comma.tf, colon.tf, semicolon.tf,`
`space.tf ` was adjusted to be half-width in 2.003, see: https://github.com/googlefonts/googlesans-flex/issues/937
